File: EosSocial/main_window.py

from gi.repository import Gdk
from gi.repository import GdkX11
from gi.repository import Gtk
from wm_inspect import WM_Inspect_MixIn

class MainWindow(Gtk.Window, WM_Inspect_MixIn):

    def __init__(self):
        Gtk.Window.__init__(self, 
                            type=Gtk.WindowType.TOPLEVEL,
                            decorated=False,
                            modal=True,
                            resizable=False,
                            skip_pager_hint=True,
                            skip_taskbar_hint=True)
        WM_Inspect_MixIn.__init__(self)

        # stick on all desktops
        self.stick()
        # do not destroy on delete event
        self.connect('delete-event', lambda w, e: self.hide_on_delete())

        # update position when workarea changes
        screen = Gdk.Screen.get_default()
        screen.connect('monitors-changed', self._on_monitors_changed)

    def _active_win_callback(self, active_xid):
        xid = GdkX11.X11Window.get_xid(self.get_window())
        if xid != active_xid:
            self.hide()

    def _on_monitors_changed(self, screen, data):
        self._ensure_position()

    def _ensure_position(self):
        screen = Gdk.Screen.get_default()
        monitor = screen.get_primary_monitor()
        workarea = screen.get_monitor_workarea(monitor)

        width = workarea.width / 3

        geometry = Gdk.Rectangle()
        geometry.x = workarea.x + workarea.width - width
        geometry.y = workarea.y
        geometry.width = width
        geometry.height = workarea.height

        self.move(geometry.x, geometry.y)
        self.set_size_request(geometry.width, geometry.height)

    def show(self):
        self._ensure_position()
        self.present()

    def toggle(self):
        if self.get_visible():
            self.hide()
        else:
            self.show()

File: EosSocial/mainWindow.js
const Gdk = imports.gi.Gdk;
const GdkX11 = imports.gi.GdkX11;
const Gtk = imports.gi.Gtk;
const Lang = imports.lang;

const WMInspect = imports.wmInspect;

const MainWindow = new Lang.Class({
    Name: 'MainWindow',
    Extends: Gtk.ApplicationWindow,

    _init: function(application) {
        this.parent({ type: Gtk.WindowType.TOPLEVEL,
                      application: application,
                      decorated: false,
                      modal: true,
                      resizable: false,
                      skip_pager_hint: true,
                      skip_taskbar_hint: true });

        this._wmInspect = new WMInspect.WMInspect();
        this._wmInspect.connect('active-window-changed', Lang.bind(this,
            this._onActiveWindowChanged));

        // stick on all desktop
        this.stick();
        // do not destroy on delete event
        this.connect('delete-event', Lang.bind(this,
            this.hide_on_delete));

        // update position when workarea changes
        let screen = Gdk.Screen.get_default();
        screen.connect('monitors-changed', Lang.bind(this,
            this._ensurePosition));
    },

    _onActiveWindowChanged: function(wmInspect, activeXid) {
        let xid = this.get_window().get_xid();
        if (xid != activeXid)
            this.hide();
    },

    _ensurePosition: function() {
        let screen = Gdk.Screen.get_default();
        let monitor = screen.get_primary_monitor();
        let workarea = screen.get_monitor_workarea(monitor);

        let width = workarea.width / 3;
        let geometry = { x: workarea.x + workarea.width - width,
                         y: workarea.y,
                         width: width,
                         height: workarea.height };

        this.move(geometry.x, geometry.y);
        this.set_size_request(geometry.width, geometry.height);
    },

    show: function() {
        this._ensurePosition();
        this.present();
    },

    toggle: function() {
        if (this.get_visible())
            this.hide();
        else
            this.show();
    }
});
